src/Serialization/Rings.jl

```diff
@@ -0,0 +1,124 @@
+function save_internal(s::SerializerState, R::Nemo.NmodRing)
+    return Dict(
+        :modulus => save_type_dispatch(s, modulus(R))
+    )
+end
+
+function load_internal(s::DeserializerState, ::Type{Nemo.NmodRing}, dict::Dict)
+    modulus = load_type_dispatch(s, UInt64, dict[:modulus])
+    return Nemo.NmodRing(modulus)
+end
+
+#elements
+function save_internal(s::SerializerState, r::nmod)
+    return Dict(
+        :parent => save_type_dispatch(s, parent(r)),
+        :class_val => save_type_dispatch(s, fmpz(r))
+    )
+end
+
+function load_internal(s::DeserializerState, ::Type{nmod}, dict::Dict)
+    parent_ring = load_type_dispatch(s, UInt64, dict[:parent])
+    class_val = load_type_dispatch(s, fmpz, dict[:class_val])
+    return parent_ring(class_val)
+end
+
+
+################################################################################
+#  Polynomial Rings
+function save_internal(s::SerializerState, R::Union{MPolyRing, PolyRing})
+    return Dict(
+        :symbols => save_type_dispatch(s, symbols(R)),
+        :base_ring => save_type_dispatch(s, base_ring(R)),
+    )
+end
+
+function load_internal(s::DeserializerState,
+                       T::Type{<: Union{MPolyRing, PolyRing}},
+                       dict::Dict)
+    base_ring = load_type_dispatch(s, dict[:base_ring], check_namespace=false)
+    symbols = load_type_dispatch(s, Vector{Symbol}, dict[:symbols])
+
+    if T <: PolyRing
+        return PolynomialRing(base_ring, symbols..., cached=false)
+    end
+
+    return PolynomialRing(base_ring, symbols, cached=false)
+end
+
+################################################################################
+# Multivariate Polynomials
+function save_internal(s::SerializerState, p::MPolyElem)
+    parent_ring = parent(p)
+    parent_ring = save_type_dispatch(s, parent_ring)
+    terms = []
+
+    for i in 1:length(p)
+        term = Dict(
+            :exponent => save_type_dispatch(s, exponent_vector(p, i)),
+            :coeff => save_type_dispatch(s, coeff(p, i))
+        )
+        push!(terms, term)
+    end
+
+    return Dict(
+        :terms => terms,
+        :parent => parent_ring,
+    )
+end
+
+function load_internal(s::DeserializerState, ::Type{<: MPolyElem}, dict::Dict)
+    R, symbols = load_type_dispatch(s, dict[:parent], check_namespace=false)
+    coeff_ring = coefficient_ring(R)
+    coeff_type = elem_type(coeff_ring)
+    polynomial = MPolyBuildCtx(R)
+
+    for term in dict[:terms]
+        c = load_type_dispatch(s, coeff_type, term[:coeff])
+        e = load_type_dispatch(s, Vector{Int}, term[:exponent])
+        push_term!(polynomial, c, e)
+    end
+    return finish(polynomial)
+end
+
+################################################################################
+# Univariate Polynomials
+function save_internal(s::SerializerState, p::PolyElem)
+    parent_ring = parent(p)
+    parent_ring = save_type_dispatch(s, parent_ring)
+
+    return Dict(
+        :parent => parent_ring,
```

Comment:
Maybe like this for consistency for other similar code?
```suggestion
    return Dict(
        :parent => save_type_dispatch(s, parent(p)),
```

Reply:
So sometimes it is necessary to save the parent before, otherwise we might have a ref that is deeper than a backref in the dict. by saving the parent before going to the dictionary you have control on this, and guarantee that the ref stays as high on the dict as possible, to not throw an error on load

Reply:
I don't understand -- as far as I can tell, my suggested change does not affect the execution order at all? I.e., `save_type_dispatch(s, parent(p))` is still executed first, before the `Dict` is created and returned